core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala

```diff
@@ -338,25 +340,39 @@ class WorkspaceService(protected val userInfo: UserInfo, val dataSource: SlickDa
     }
   }
 
-  def updateWorkspace(workspaceName: WorkspaceName, operations: Seq[AttributeUpdateOperation]): Future[PerRequestMessage] =
-    withAttributeNamespaceCheck(operations.map(_.name)) {
-      dataSource.inTransaction { dataAccess =>
-        withWorkspaceContextAndPermissions(workspaceName, WorkspaceAccessLevels.Write, dataAccess) { workspaceContext =>
-          val updateAction = Try {
-            val updatedWorkspace = applyOperationsToWorkspace(workspaceContext.workspace, operations)
-            dataAccess.workspaceQuery.save(updatedWorkspace)
-          } match {
-            case Success(result) => result
-            case Failure(e: AttributeUpdateOperationException) =>
-              DBIO.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.BadRequest, s"Unable to update ${workspaceName}", ErrorReport(e))))
-            case Failure(regrets) => DBIO.failed(regrets)
-          }
-          updateAction.map { savedWorkspace =>
-            RequestComplete(StatusCodes.OK, savedWorkspace)
-          }
+  def updateLibraryAttributes(workspaceName: WorkspaceName, operations: Seq[AttributeUpdateOperation]): Future[PerRequestMessage] = {
+    withLibraryAttributeNamespaceCheck(operations.map(_.name)) {
+      getWorkspaceContext(workspaceName) flatMap { ctx =>
+        withLibraryPermissions(ctx.workspace, operations, dataSource, userInfo, getMaximumAccessLevel _) {
+          updateWorkspace(workspaceName, operations, ctx)
         }
       }
     }
+  }
+
+  def updateWorkspace(workspaceName: WorkspaceName, operations: Seq[AttributeUpdateOperation]): Future[PerRequestMessage] = {
+    withAttributeNamespaceCheck(operations.map(_.name)) {
+    getWorkspaceContextAndPermissions(workspaceName, WorkspaceAccessLevels.Write) flatMap { ctx =>
+        updateWorkspace(workspaceName, operations, ctx)
+      }
+    }
+  }
+
+  private def updateWorkspace(workspaceName: WorkspaceName, operations: Seq[AttributeUpdateOperation], workspaceContext: SlickWorkspaceContext): Future[PerRequestMessage] =
+    dataSource.inTransaction { dataAccess =>
+      val updateAction = Try {
+        val updatedWorkspace = applyOperationsToWorkspace(workspaceContext.workspace, operations)
+        dataAccess.workspaceQuery.save(updatedWorkspace)
+      } match {
+        case Success(result) => result
+        case Failure(e: AttributeUpdateOperationException) =>
+          DBIO.failed(new RawlsExceptionWithErrorReport(errorReport = ErrorReport(StatusCodes.BadRequest, s"Unable to update ${workspaceName}", ErrorReport(e))))
+        case Failure(regrets) => DBIO.failed(regrets)
+      }
+      updateAction.map { savedWorkspace =>
+        RequestComplete(StatusCodes.OK, savedWorkspace)
+      }
+    }
```

Comment:
Hmm ... in the `updateLibraryAttributes`/`updateWorkspace` refactor - which I like from a modularity perspective - the db transaction boundaries changed. Now, it's theoretically possible that the workspace changed between the time we get it and the time we update it. Consider moving the transaction boundary back out of the private method and into the two calling methods?